### PR TITLE
fast hex functions

### PR DESCRIPTION
**Pull request by zbtzbtzbt**
## Proposed changes

原谅我标题党了。。。我感觉至少比boost里面的hex快，主要原因是：减少了虚函数调用，确定输入的类型就是char*了。

## Types of changes

What types of changes does your code introduce to Doris?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Code refactor (Modify the code structure, format the code, etc...)
- [ ] Optimization. Including functional usability improvements and performance improvements.
- [ ] Dependency. Such as changes related to third-party components.
- [ ] Other.

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [ ] I have created an issue on (Fix #ISSUE) and described the bug/feature there in detail
- [ ] Compiling and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If these changes need document changes, I have updated the document
- [ ] Any dependent changes have been merged

## Further comments

If this is a relatively large or complex change, kick off the discussion at dev@doris.apache.org by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: be/src/vec/functions/function_string.cpp [hunk 1]
```diff
@@ -243,6 +243,156 @@ struct StringFunctionImpl {
...
+        //check and decode one character at the same time
+        // character in ( ['0','9'], ['a','f'], ['A','F'] ), return 'NULL' like mysql dose.
+        for (auto i = 0, dst_index = 0; i < src_len; i += 2, dst_index++) {
+            char c = 0;
+            // combine src_str two character into dst_str one character , bool param means weather multiple 16.
+            bool left_4bits_flag = check_and_decode_one(c, *(src_str + i), true);
+            bool right_4bits_flag = check_and_decode_one(c, *(src_str + i + 1), false);
+            
+            if (!left_4bits_flag || !right_4bits_flag) {
+                *dst_str = 'N';
```
Comment (BiteTheDDDDt) [inline be/src/vec/functions/function_string.cpp:348]: I think returning null means marking a NullMap in a nullable column, not a string with a value of `NULL`.

Comment (zbtzbtzbt) [reply be/src/vec/functions/function_string.cpp:348]: ok, i will change it.

File: be/src/vec/functions/function_string.cpp [hunk 2]
```diff
@@ -243,6 +243,156 @@ struct StringFunctionImpl {
     }
 };
 
+struct NameHex {
+    static constexpr auto name = "hex";
+};
+struct HexImpl {
+    static void fast_hex_encode(const char* src_str, char* dst_str,
+                                const char* hexDigits = "0123456789ABCDEF") {
```
Comment (HappenLee) [inline be/src/vec/functions/function_string.cpp:251]: 咱们应该是下划线命名法

**Review by HappenLee — COMMENTED** [review]
请用测试证明一下他足够“fast”~~

**Comment by zbtzbtzbt** [comment]
在实现hex(int)时发现，函数参数是columnInt64会导致BE挂掉，改成columnInt32不会。猜测resize(data.size())类型大小或者内存对齐有关，不知道这个问题是不是普遍存在的。

**Comment by HappenLee** [comment]
> 在实现hex(int)时发现，函数参数是columnInt64会导致BE挂掉，改成columnInt32不会。问题出在resize(data.size())这一句，猜测和data.size()类型大小或者resize内存对齐有关，不知道这个问题是不是普遍存在的。

这个问题很奇怪啊，hex函数本身的参数就是bigint，应该不会导致类似的问题的

**Pull request closed without merging**